### DISCUSSION TITLE
Convolutional Neural Network in Python Not Learning Effectively: Issues with Backpropagation

### DISCUSSION THREAD

**Discussion by duk7292** [community/community#65112] [Programming Help]
Hello everyone,

I've been working on implementing a Convolutional Neural Network (CNN) from scratch in Python to recognize individual letters. My main goal is to deepen my understanding of how CNNs operate. While I've set up the entire structure and it does generate some output, the network doesn't seem to be learning effectively.

Specific Issues:

1.Error mean/Variance: 

During backpropagation, I'm observing that the mean and variance of the error for each layer are behaving erratically. Here are some sample metrics:

```
mean: 4.503609024170557e-19 of <class 'utils.SoftmaxLayer'>
var: 8.736755909119868e-05 of <class 'utils.SoftmaxLayer'>
mean: 0.00016175699266982586 of <class 'utils.DenseLayer'>
var: 3.84936955627577e-05 of <class 'utils.DenseLayer'>
mean: -5.232296890610626e-05 of <class 'utils.ReLuLayer'>
var: 2.053791717515726e-05 of <class 'utils.ReLuLayer'>
mean: -9.927402112160702e-07 of <class 'utils.DenseLayer'>
var: 1.0367016318542532e-07 of <class 'utils.DenseLayer'>
mean: -9.927402112160702e-07 of <class 'utils.FlattenLayer'>
var: 1.0367016318542532e-07 of <class 'utils.FlattenLayer'>
mean: -2.4818505280401765e-07 of <class 'utils.MaxPooling2D'>
var: 2.5917725583817643e-08 of <class 'utils.MaxPooling2D'>
mean: -2.919212928001479e-07 of <class 'utils.ReLuLayer'>
var: 1.831728681055942e-08 of <class 'utils.ReLuLayer'>
mean: -1.5161430918119097e-05 of <class 'utils.ConvolutionLayer'>
var: 8.692331085156676e-07 of <class 'utils.ConvolutionLayer'>
mean: -3.7903577295297743e-06 of <class 'utils.MaxPooling2D'>
var: 2.173513775640703e-07 of <class 'utils.MaxPooling2D'>
mean: 2.5000857230143796e-06 of <class 'utils.ReLuLayer'>
var: 1.3191464399833077e-07 of <class 'utils.ReLuLayer'>
mean: -3.824830351339615e-05 of <class 'utils.ConvolutionLayer'>
var: 2.2049322150705984e-06 of <class 'utils.ConvolutionLayer'>
```

2.Limited Class Recognition: 

When I train the network on a limited set of letters like 'a', 'b', and 'c', it seems to specialize in recognizing only those letters but fails to differentiate between them.

I've spent the last week rigorously checking for errors in my code but haven't found any glaring issues. I know it's a lot to ask for, but I would be incredibly grateful if someone could review the code to spot any obvious issues or provide any insights into what might be going wrong. At this point, I really don't know what to do.

so here is my code:

main:
```
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import utils as ut

learning_rate = 0.001


def main():
    dataRaw = pd.read_csv('balanced_data.csv').to_numpy().astype(float)
    np.random.shuffle(dataRaw)
    
    data_Normalized = ut.normalizeData(dataRaw)
    data_Normalized = np.array(data_Normalized)

    labels = data_Normalized[:, 0].astype(int)
    images_1d = data_Normalized[:, 1:]
    
    image_size = 28
    images_2d = images_1d.reshape(-1, image_size, image_size)

    labels_train, labels_test, images_train, images_test = train_test_split(labels, images_2d, test_size=0.1)

    layers = [
        ut.ConvolutionLayer(32,3,learning_rate),
        ut.ReLuLayer(),
        ut.MaxPooling2D(2),
        ut.ConvolutionLayer(32,3,learning_rate),
        ut.ReLuLayer(),
        ut.MaxPooling2D(2),
        ut.FlattenLayer(),
        ut.DenseLayer(126,50176,learning_rate),
        ut.ReLuLayer(),
        ut.DenseLayer(26,126,learning_rate),
        ut.SoftmaxLayer()
            ]
    def train_in_batches(images_train, labels_train, layers, batch_size=5):
        num_batches = len(images_train) // batch_size
        accuracies = []
        for batch_idx in range(num_batches):
            batch_predictions = []
            batch_labels = []

            if(batch_idx % 300 == 0):
                layers[0].learning_rate/=10
                layers[3].learning_rate/=10
                layers[7].learning_rate/=10
                layers[9].learning_rate/=10

         
            start_idx = batch_idx * batch_size
            end_idx = start_idx + batch_size
            batch_Errors = []
            
            for i, (image, label) in enumerate(zip(images_train[start_idx:end_idx], labels_train[start_idx:end_idx])):

                output = ut.CNN_forward(image, layers)
                
                label_arr = np.zeros(len(output))
                label_arr[label] = 1
                
                
                batch_predictions.append(output)
                batch_labels.append(label)

                error =(output- label_arr)  
              
                batch_Errors.append(error)

            stacked_error = np.stack(batch_Errors)

           
            avg_Error =  np.mean(stacked_error, axis=0)
            
            ut.CNN_backward(avg_Error, layers)
            
            accuracy = ut.calculate_accuracy(batch_predictions, batch_labels)
            accuracies.append(accuracy)
            
        
            batch_predictions = np.array(batch_predictions)
            batch_labels = np.array(batch_labels)
  
            print(f"Genauigkeit für Batch {batch_idx + 1}: {accuracy:.2f}%\nDurchschnittliche Genauigkeit über alle Batches: { np.mean(accuracies):.2f}%\nHöste Genauigkeit bis Jetzt {accuracies[np.argmax(accuracies)]:.2f}% bei batch {np.argmax(accuracies)}")


    train_in_batches(images_train, labels_train, layers) 
        
        
        

if __name__ == "__main__":
    main()
```
utils:

```
import numpy as np


class ConvolutionLayer:
    def __init__(self,kernel_num,kernel_size,learning_rate) -> None:
        self.learning_rate = learning_rate
        self.kernel_num = kernel_num
        self.kernel_size = kernel_size
        self.kernels = np.random.randn(kernel_num, kernel_size, kernel_size) * np.sqrt(2. / (kernel_size * kernel_size))


        self.biases = np.random.randn(kernel_num) * 0.01


        self.input = None
    def forward(self,images):
        self.input = images
        convolution_output  = []
        
        for kernel,bias in zip(self.kernels,self.biases):
            for image in images:
                kernel_conv = conv2d_same(image,kernel)
                kernel_conv += bias

                convolution_output.append(kernel_conv) 
        
        return np.array(convolution_output)
    def backward(self, errors):
        dKernels = np.zeros_like(self.kernels)
        dBiases = np.zeros_like(self.biases)
        dInputs = np.zeros_like(self.input)
        
        
        for i, (kernel, bias) in enumerate(zip(self.kernels, self.biases)):
            for j, (image, error) in enumerate(zip(self.input, errors)):

                dKernels[i] += conv2d_valid(image, error)
                

                dBiases[i] += np.sum(error)
                

                rotated_kernel = np.rot90(kernel, 2)
                dInputs[j] += conv2d_same(error, rotated_kernel)
        self.kernels -= self.learning_rate * dKernels
        self.biases -= self.learning_rate * dBiases
        return dInputs

class MaxPooling2D:
    def __init__(self,kernel_size) -> None:
        self.kernel_size = kernel_size
        self.input = None
        self.max_mask = None
    def forward(self,images):
        self.input = images
        pooled_output = np.zeros((len(images), *maxpooling2d(images[0], 2).shape))
        for idx, image in enumerate(images):
            pooled_output[idx] = maxpooling2d(image, 2)
        
        

        enlarged_pooled_output = pooled_output.repeat(1, axis=0).repeat(self.kernel_size, axis=1).repeat(self.kernel_size, axis=2)
        self.max_mask = (images == enlarged_pooled_output).astype(int)
        correct_max_mask_sum = int((images.shape[0]*(images[0].shape[0]*images[0].shape[1]))/self.kernel_size**2)
        # if there are multiple max values for one patch
        if(np.sum(self.max_mask) > correct_max_mask_sum):
            self.max_mask =  correct_max_mask(self.max_mask,self.kernel_size)
        
        return pooled_output

    def backward(self,error):
        dinput = np.zeros(self.input.shape)
        for n in range(error.shape[0]):
            for i in range(error.shape[1]):
                for j in range(error.shape[2]):
                    dinput[n, i*self.kernel_size:(i+1)*self.kernel_size, j*self.kernel_size:(j+1)*self.kernel_size] = error[n, i, j] * self.max_mask[n, i*self.kernel_size:(i+1)*self.kernel_size, j*self.kernel_size:(j+1)*self.kernel_size]        
        return dinput
    
class DenseLayer:
    def __init__(self,output_size,input_size,learning_rate):
        self.learning_rate = learning_rate
        self.output_size = output_size
        self.input_size = input_size

        self.weights = np.random.randn(self.output_size, self.input_size) * np.sqrt(2. / self.input_size)


        self.bias = np.random.randn(self.output_size) * 0.01


    
        self.input = None
    def forward(self,input):
        
        self.input = input
        output = np.dot(self.weights,input)+(self.bias.reshape(self.output_size))
        
        return output

    def backward(self,error):
        error = error.reshape(-1, 1)
        input_reshaped = self.input.reshape(1, -1)
        dW = np.dot(error, input_reshaped)
        db = np.sum(error, axis=1)
        dA_prev = np.dot(self.weights.T, error)
        
        self.bias -= self.learning_rate * db
        self.weights -= self.learning_rate *  dW
        return dA_prev.ravel()

class FlattenLayer:
    def __init__(self) -> None:
        self.input = None
    def forward(self,input):
        self.input = input
        return input.flatten()
     
    def backward(self,error):
        return error.reshape(self.input.shape)

class ReLuLayer:
    def __init__(self) -> None:
        self.input = None
    def forward(self,input):
        self.input = input
        return np.maximum(0, input)
    
    def backward(self, error):
        
        error[self.input < 0] = 0 
        return  error

    


class SoftmaxLayer:
    def __init__(self, axis=-1) -> None:
        self.output = None
        self.axis = axis

    def forward(self, input):
        exps = np.exp(input - np.max(input, axis=self.axis, keepdims=True))  
        self.output = exps / np.sum(exps, axis=self.axis, keepdims=True)
        return self.output

    def backward(self, error):
        grad = self.output * (error - np.sum(error * self.output, axis=self.axis, keepdims=True))
        return grad


def CNN_forward(image,layers):
    output = image.reshape(1, image.shape[0], image.shape[1])

    for layer in layers:     
        output = layer.forward(output)
 
    return (output)
def CNN_backward(error_output,layers):
    error =error_output
    for layer in reversed(layers):
        error = layer.backward(error)


def normalizeData(data):    
    data[:, 1:] = data[:, 1:] / 255.0
    return  data 


def calculate_accuracy(predictions, labels):
    predicted_labels = np.argmax(predictions, axis=1)
    correct_predictions = np.sum(predicted_labels == labels)
    accuracy = (correct_predictions / len(labels)) * 100
    return accuracy

def conv2d_valid(input_array, kernel, output_size=(3, 3)):
    x, y = input_array.shape
    a, b = kernel.shape

    output_x, output_y = output_size

    output = np.zeros((output_x, output_y))

    for i in range(output_x):
        for j in range(output_y):
            if i + a <= x and j + b <= y:
                output[i, j] = np.sum(input_array[i:i+a, j:j+b] * kernel)
            else:
                temp_array = np.zeros((a, b))
                temp_array[:min(x-i, a), :min(y-j, b)] = input_array[i:min(i+a, x), j:min(j+b, y)]
                output[i, j] = np.sum(temp_array * kernel)

    return output



def conv2d_same(input_array, kernel):
    x, y = input_array.shape
    a, b = kernel.shape

    pad_x = a // 2
    pad_y = b // 2

    padded_input = np.pad(input_array, ((pad_x, pad_x), (pad_y, pad_y)), mode='constant')

    output = np.zeros_like(input_array)

    for i in range(x):
        for j in range(y):
            output[i, j] = np.sum(padded_input[i:i+a, j:j+b] * kernel)

    return output

def maxpooling2d(input_matrix, kernel_size):
    rows, cols = input_matrix.shape
    
    pooled_rows = rows // kernel_size
    pooled_cols = cols // kernel_size
    output_matrix = np.zeros((pooled_rows, pooled_cols))
    for i in range(pooled_rows):
        for j in range(pooled_cols):
            output_matrix[i , j ] = np.max(input_matrix[(kernel_size*i):(kernel_size*i)+kernel_size, (kernel_size*j):(kernel_size*j)+kernel_size])
    
    return output_matrix

def correct_max_mask(max_masks, Kernel_size):
    correct_max_mask = np.zeros_like(max_masks)
    for i, max_mask in enumerate(max_masks):
        for y in range(max_mask.shape[0]):
            for x in range(max_mask.shape[1]):
                s_x = x // Kernel_size
                s_y = y // Kernel_size
                patch = correct_max_mask[i, s_y*Kernel_size:s_y*Kernel_size+Kernel_size, s_x*Kernel_size:s_x*Kernel_size+Kernel_size]
                patch_sum = np.sum(patch)
                if patch_sum == 0:
                    correct_max_mask[i, y, x] = max_mask[y, x]
    return correct_max_mask

```

here are some things that i tried

Omitting Softmax Layer: 

Given that the mean error was smallest in the Softmax layer, I decided to remove it temporarily to see if it would resolve the problem. However, the output for the mean and variance afterwards was inconclusive, and I couldn't derive much from it. Here's what the output looked like:

```
mean: 0.15940985428873375 of <class 'utils.ReLuLayer'>
var: 0.20957548527481673 of <class 'utils.ReLuLayer'>
mean: 0.0003047370398286636 of <class 'utils.DenseLayer'>
var: 0.001168835152196895 of <class 'utils.DenseLayer'>
mean: 0.0003047370398286636 of <class 'utils.FlattenLayer'>
var: 0.001168835152196895 of <class 'utils.FlattenLayer'>
mean: 7.618425995716592e-05 of <class 'utils.MaxPooling2D'>
var: 0.00029222620017361946 of <class 'utils.MaxPooling2D'>
mean: 6.349381565703456e-05 of <class 'utils.ReLuLayer'>
var: 0.000191664771879905 of <class 'utils.ReLuLayer'>
mean: -0.00016786510964692848 of <class 'utils.ConvolutionLayer'>
var: 0.012032811595479576 of <class 'utils.ConvolutionLayer'>
mean: -4.19662774117321e-05 of <class 'utils.MaxPooling2D'>
var: 0.0030082081823752127 of <class 'utils.MaxPooling2D'>
mean: 0.00014592454642264617 of <class 'utils.ReLuLayer'>
var: 0.0014236958374105224 of <class 'utils.ReLuLayer'>
mean: -0.004472683933071659 of <class 'utils.ConvolutionLayer'>
var: 0.10245428401793283 of <class 'utils.ConvolutionLayer'>
```

Changing Weight Initialization to Kaiming:

 I switched the weight initialization method to Kaiming, hoping that it would stabilize the variance. Unfortunately, this did not lead to any significant changes.

Minor Bug Fixes:

 I've resolved several minor issues in the implementation, but there are too many to list here.

Despite these attempts, I'm still facing the same problems.

Again any help would be awesome, and thanks a ton for taking the time to read my post!

**Comment by arafajs** [comment]
- The network seems to be overfitting on the small training set. Try more aggressive regularization techniques like dropout layers.
- Use data augmentation (rotations, shifts, zooms, etc) to expand the limited training data.
- Try training for fewer epochs - it may be overfitting after just a couple epochs on the small dataset.
- Try smaller learning rates like 0.0001 or 0.00001 and see if it stabilizes.
- Add batch normalization layers which can help stabilize training.
- Monitor the weight and bias values during training - if they are becoming very large, it can cause the strange error behavior.

**Reply by duk7292** [reply]
thanks i will try that

**Reply by duk7292** [reply]
but one more question why do the error values become extremely low because if i use smaller learning rate its basically 0? This happens whether or not I include a softmax layer. What could be the reason for this?
